## Data Cleaning and Preprocessing
In this section, we go through the data cleaning and preprocessing steps taken to prepare the data in this project.


### Data Retrieval
- We load the modules required for this report.
- We read in the dataset from diamonds.csv.
- We display 10 randomly sampled rows from the report to get an idea of what our data looks like.

In [1]:
###Loading modules###
import numpy as np
import pandas as pd

#used to get the csv from the repository
import requests

#used to generate tables to check data
from IPython.display import display, HTML

###Reading dataset as a dataframe###
df_name = 'diamonds.csv'
diamonds_df = pd.read_csv(df_name)

###Displaying 10 random rows###
diamonds_df.sample(10, random_state=405)

carat        cut color clarity  depth  table     x     y     z  price
36187   0.41  Very Good     F     VS1   63.0   57.0  4.66  4.70  2.95    931
51096   0.71  Very Good     G     VS2   62.3   63.0  5.70  5.64  3.53   2335
2414    0.87  Very Good     H     SI1   62.2   57.0  6.05  6.11  3.78   3180
2136    0.71      Ideal     G     VS1   61.8   56.1  5.72  5.74  3.54   3123
47156   0.45      Ideal     F    VVS1   61.8   55.0  4.92  4.95  3.05   1833
11719   0.34      Ideal     E     SI1   61.4   54.0  4.50  4.53  2.77    596
51509   0.74    Premium     F     SI2   59.2   60.0  5.96  5.90  3.51   2383
27566   2.40      Ideal     H     SI2   62.0   55.0  8.57  8.60  5.32  18374
33746   0.31      Ideal     F    VVS1   62.0   55.0  4.34  4.37  2.70    840
35068   0.32  Very Good     F    VVS1   61.5   56.0  4.38  4.40  2.70    886

### Data Cleaning
- Drop irrelevant columns from our dataset.
- Verify that all columns are labelled properly.
- Check for missing values and remove rows that have missing values.
- Random sampling of 5k rows?

First, we will display all column names too see whether any columns need to be dropped.

In [2]:
#displays all columns
diamonds_df.columns

Index(['carat', 'cut', 'color', 'clarity', 'depth', 'table', 'x', 'y', 'z',
       'price'],
      dtype='object')

`x`, `y` and `z` represent the lengths of each dimension of the diamond which is redundant as `depth` and `table` are provided and are the industry standard way of measuring diamonds size. Because of this we will drop `x`, `y` and `z`.

In [3]:
#Dropping x y z columns
diamonds_df = diamonds_df.drop(columns=['x', 'y', 'z'])
#Display remaining columns
diamonds_df.sample(10, random_state = 405)

carat        cut color clarity  depth  table  price
36187   0.41  Very Good     F     VS1   63.0   57.0    931
51096   0.71  Very Good     G     VS2   62.3   63.0   2335
2414    0.87  Very Good     H     SI1   62.2   57.0   3180
2136    0.71      Ideal     G     VS1   61.8   56.1   3123
47156   0.45      Ideal     F    VVS1   61.8   55.0   1833
11719   0.34      Ideal     E     SI1   61.4   54.0    596
51509   0.74    Premium     F     SI2   59.2   60.0   2383
27566   2.40      Ideal     H     SI2   62.0   55.0  18374
33746   0.31      Ideal     F    VVS1   62.0   55.0    840
35068   0.32  Very Good     F    VVS1   61.5   56.0    886

Fortunately, all our column names are already correct we can now check to make sure all of our columns data types are the types we intended.

In [4]:
# Check for data types
print(f"Shape of the dataset = {diamonds_df.shape} \n")
print(f"Data types are below where 'object' indicates a string type: ")
print(diamonds_df.dtypes)

Shape of the dataset = (53940, 7) 

Data types are below where 'object' indicates a string type: 
carat      float64
cut         object
color       object
clarity     object
depth      float64
table      float64
price        int64
dtype: object


Next, we will check all our categorical columns for their unique values and ensure that all entries are valid.

In [5]:
#all cols that contain catergorical data (object = string)
catCols = diamonds_df.columns[diamonds_df.dtypes==object].tolist()

for col in catCols:
    print(f"All unique vals in {col}")
    print(f"{diamonds_df[col].unique()}\n")

All unique vals in cut
['Ideal' 'Premium' 'Good' 'Very Good' 'Fair']

All unique vals in color
['E' 'I' 'J' 'H' 'F' 'G' 'D']

All unique vals in clarity
['SI2' 'SI1' 'VS1' 'VS2' 'VVS2' 'VVS1' 'I1' 'IF']



All the unique values are valid and so we can move on.

Next, we will check for outliers in our numerical data by generating a summary of our statistics. From looking at this we can determine that there are no outliers present that we need to act on.

In [6]:
display(HTML('<b>Table: Summary of numerical features</b>'))
diamonds_df.describe(include=['int64','float64']).T

count         mean          std    min    25%     50%      75%  \
carat  53940.0     0.797940     0.474011    0.2    0.4     0.7     1.04   
depth  53940.0    61.749405     1.432621   43.0   61.0    61.8    62.50   
table  53940.0    57.457184     2.234491   43.0   56.0    57.0    59.00   
price  53940.0  3932.799722  3989.439738  326.0  950.0  2401.0  5324.25   

            max  
carat      5.01  
depth     79.00  
table     95.00  
price  18823.00

Missing values are checked by displaying the number of missing values in every column. We observe that fortunately there are no missing values in any column so we can continue.

In [7]:
#get num of missing values in each column
print(f"\nColumn:\t   Num missing:")
print(diamonds_df.isnull().sum())


Column:	   Num missing:
carat      0
cut        0
color      0
clarity    0
depth      0
table      0
price      0
dtype: int64


### Random Sampling
We have 53940 clean rows so we will take a random sample of 5000 rows. Then we will display a random sample of 10 from the 5000 rows.

In [8]:
diamonds_df = diamonds_df.sample(5000, random_state = 405)
diamonds_df.sample(10, random_state = 405)

carat        cut color clarity  depth  table  price
14074   1.03      Ideal     E     SI1   62.0   53.0   5710
32009   0.31  Very Good     F    VVS1   59.2   62.0    777
40757   0.40      Ideal     F    VVS2   62.8   56.0   1163
21811   1.66    Premium     J    VVS2   62.6   59.0   9882
32701   0.31      Ideal     F     VS2   61.6   55.0    802
50018   0.70  Very Good     F     SI1   59.8   60.0   2196
17483   1.10    Premium     G     VS2   59.4   60.0   7022
24064   1.70      Ideal     I     VS1   60.1   57.0  12209
36964   0.41    Premium     G     VS1   62.5   59.0    961
45110   0.54  Very Good     D     SI1   60.3   61.0   1650